proposal/abc225666/README.md

```diff
@@ -0,0 +1,75 @@
+# Agar backend api
+
+[![hackmd-github-sync-badge](https://hackmd.io/RYp6o8DmTr2QNHb7f-hbCg/badge)](https://hackmd.io/RYp6o8DmTr2QNHb7f-hbCg)
+
+
+## Basic information
+Github repository: https://github.com/abc225666/agar-backend-api
+
+## Problem to solve
+Agar is mini game produced by agar.io. People control a circle to competitive with other people. The larger your circle is, the more point you get and your circle is more powerful.
+
+A circle can eat other circles to grow up if it is large enough. Circle can also split it self, send circle out for other circle. With these skill, players need to compete or corporate with others to get high points.
+
+This project would like to supply lots of apis about Agar, like calculating circles growth, players' interaction in a map. It will involve lots of numerical techniques like circle area calculation, coordinate after moving, and scanning for circle overlapping to detect which circle will be eaten. 
+
+## Prospective user
+Users who want to build a Agar server could use these apis to decrease the workload in calculate circle's interaction.
+
+## System architecture
+#### Class `Circle`
+Contains all data related to this circle, including radius, position, the user belongs to.
+
+#### Class `User`
+Contains user's data
+
+#### Class `Map`
+Contains map's information, might include maximum area of map.
+
+#### Program Flow
+When a user join the game, a circle should be assigned to him.  
+After every user action, I might scan all possible changes and calculate new data for every circle.
+
+## API description
+I will design detailed API when doing project. At this time, I only simply describe what API will be there.
+- Function when a user joining the game.
+- When a circle eat other circle, calculate new area size of the circle.
+- Moving function of a circle.
+- Checking circle overlapping, if it matches the condition, one of circles will be eaten.
+- Split circle, it will become **2n** circles.
+- Function of sending out circles, it will make the circle become small, and a new freedom circle will be send out to map.
+
+## Engineering infrastructure
+- Public APIs (Python)
+  - Supply api that user can call directly to get calculate result.
+- Private APIs (C++)
+  - Circle growth calculation
+  - Circle eat others
+  - Circle split
+  - Circle send small circle into map
+  - Map related function, generate freedom circles, etc.
+
+** I supply the general design first. The detailed infrastructure will continuously add when developing.
```

Comment:
It's good to have the design evolve along with the implementation.  We should keep updating the document/plan.